File: python/questions/minimumCost.py
from typing import List
from collections import deque

class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        nn = {}
        for s, e, c in zip(original, changed, cost):
            if s not in nn:
                nn[s] = {}
                nn[s][s] = 0
            nn[s][e] = min(nn[s].get(e, float('inf')), c)
        gdone = set()
        for s in nn:
            gdone.add(s)
            csc = nn[s]
            cl = deque(nn[s].keys())
            while cl:
                cll = len(cl)
                for _ in range(cll):
                    h = cl.popleft()
                    phd = nn.get(h, {})
                    for hd, cc in phd.items():
                        hdc = csc[h] + cc
                        if hdc < csc.get(hd, float('inf')):
                            csc[hd] = hdc
                            if h not in gdone:
                                cl.append(hd)
        ans = 0
        for p in range(len(source)):
            if source[p] != target[p]:
                csc = nn.get(source[p], None)
                if csc is None:
                    return -1
                cc = csc.get(target[p], float('inf'))
                if cc == float('inf'):
                    return -1
                ans += cc
        return ans

File: cpp/questions/minimumCost.cpp
#include <string>
#include <vector>
#include <climits>

using namespace std;

class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        std::vector<std::vector<int>> way(26, std::vector<int>(26, INT_MAX));
        for (int p=0; p<original.size(); p++) {
            int r = original[p] - 'a';
            int c = changed[p] - 'a';
            way[r][c] = std::min(way[r][c], cost[p]);
        }
        for (int p=0; p<26; p++) {
            way[p][p] = 0;
        }
        for (int f=0; f<26; f++) {
            for (int r=0; r<26; r++) {
                for (int c=0; c<26; c++) {
                    long long jc = static_cast<long long>(way[r][f]) + static_cast<long long>(way[f][c]);
                    if (jc > INT_MAX) {
                        continue;
                    }
                    way[r][c] = std::min(way[r][c], static_cast<int>(jc));
                }
            }
        }
        long long ans = 0;
        for (int p=0; p<source.size(); p++) {
            if (source[p] != target[p]) {
                int cc = way[source[p]-'a'][target[p]-'a'];
                if (cc == INT_MAX) {
                    return -1;
                }
                ans += cc;
            }
        }
        return ans;
    }
};
